pkg/scheduler/core/generic_scheduler.go

```diff
@@ -51,22 +51,71 @@ func (g *genericScheduler) Schedule(ctx context.Context, placement *policyv1alph
+	// 1. calculate reserved clusters and candidate clusters
+	// 2. prioritize candidate clusters
+	// 3. select target clusters from (reserved clusters + candidate clusters)
+	// 4. assign replicas
+	readyClusters, unreadyClusters, unjoinedClusters := clusterInfoSnapshot.InspectClusters()
+	bindClusters := util.ConvertToClusterNames(spec.Clusters)
+
+	reservedClusters := bindClusters.Intersection(readyClusters)
+	// Remove reserved clusters that don't fit the placement
+	for clusterName := range reservedClusters {
+		clusterObj := clusterInfoSnapshot.GetCluster(clusterName)
+		resMap := g.scheduleFramework.RunFilterPlugins(ctx, placement, &spec.Resource, clusterObj.Cluster())
+		res := resMap.Merge()
+		if !res.IsSuccess() {
+			klog.V(4).Infof("cluster %q is not fit", clusterName)
+			reservedClusters.Delete(clusterName)
+		}
+	}
+
+	availableClusters := readyClusters.Difference(bindClusters)
+	candidateClusters, err := g.findClustersThatFit(ctx, g.scheduleFramework, placement, &spec.Resource, clusterInfoSnapshot, availableClusters)
 	if err != nil {
 		return result, fmt.Errorf("failed to findClustersThatFit: %v", err)
 	}
-	if len(feasibleClusters) == 0 {
-		return result, fmt.Errorf("no clusters fit")
+
+	if len(candidateClusters) == 0 {
+		// just warn
+		klog.Warningf("There is no candidate clusters")
+	} else {
+		klog.V(4).Infof("Candidate clusters found: %v", candidateClusters)
 	}
-	klog.V(4).Infof("feasible clusters found: %v", feasibleClusters)
 
-	clustersScore, err := g.prioritizeClusters(ctx, g.scheduleFramework, placement, feasibleClusters)
+	unreadyTargetClusters := bindClusters.Intersection(unreadyClusters)
+	unjoinedTargetClusters := bindClusters.Intersection(unjoinedClusters)
+	if unreadyTargetClusters.Len() > 0 {
+		if !features.FeatureGate.Enabled(features.Failover) {
+			// reserve unready clusters if failover is disabled
+			reservedClusters = reservedClusters.Union(unreadyTargetClusters)
+		} else if len(candidateClusters) < unreadyTargetClusters.Len()+unjoinedTargetClusters.Len() {
+			// *don't evict unready clusters in binding.spec.clusters if there are insufficient available clusters*
+			// for ReplicaSchedulingTypeDivided, we will try to migrate replicas to the other health clusters
+			if placement.ReplicaScheduling == nil || placement.ReplicaScheduling.ReplicaSchedulingType == policyv1alpha1.ReplicaSchedulingTypeDuplicated {
+				// reserve unready clusters
+				klog.Warningf("reserve unready clusters due to insufficient available clusters")
+
+				// this is usually 1, which could be more than 1 in some corner cases
+				reservedNum := unreadyTargetClusters.Len() + unjoinedTargetClusters.Len() - len(candidateClusters)
+				if reservedNum > unreadyTargetClusters.Len() {
+					reservedNum = unreadyTargetClusters.Len()
+				}
+				reservedClusters = reservedClusters.Insert(unreadyTargetClusters.List()[:reservedNum]...)
```

Comment:
It is a little complicated for me to understand this logic. Do you want to remove unjoin clusters from origin schedule result and keep unready clusters?

Reply:
Yes

Reply:
This is a very unusual case that unjoin and unready events happen at the same time, but we have to consider that corner case though